torrent/networking.py

```diff
@@ -0,0 +1,188 @@
+import asyncio
+import struct
+import logging
+from enum import Enum
+
+logger = logging.getLogger(__name__)
+BLOCK_SIZE = 2**14
+CONST_PROTOCOL = 'BitTorrent protocol'.encode('utf-8')
+
+
+#out messages
+def generate_keep_alive():
+    return b'0000'
+
+
+def generate_interested_message():
+    return struct.pack('>IB', 1, 2)
+
+
+def generate_handshake(info_hash, peer_id):
+    peer_id = peer_id.encode('utf-8')
+    return struct.pack('>B19s8x20s20s', 19, CONST_PROTOCOL, info_hash, peer_id)
+
+
+def generate_request_message(piece_index, block_offset, block_size):
+    return struct.pack('>IBIII', 13, 6, piece_index, block_offset, block_size)
+
+
+#responses
+def valid_res_hand_shake(res):
+    try:
+        num, protocol, _, _ = struct.unpack('>B19s8x20s20s', res)
+    except struct.error:
+        logger.debug("message got empty")
+        return False
+    if num != 19 and protocl == CONST_PROTOCOL:
+        logger.warning('you didnt got handshake')
+        return False
+    return True
+    
+
+def get_message_type(peer_response):
+    """
+        False empty
+       -1 - keep alive
+        0 - choke
+        1 - unchoke
+        2 - interested
+        3 - uninterested
+        4 - have
+        5 - bitfield (what i got my brother)
+        6 - request
+        7 - piece
+        8 - cancel
+    """ 
+    #empty
+    if peer_response == b'':
+        logger.debug('we got empty response')
+    #keep alive
+    elif len(peer_response) == 4 and peer_response == b'0000':
+        return -1, 0
+    #message type
+    elif peer_response[4] in range(0,8):
+        try:
+            return struct.unpack(">IB", peer_response)
+        except struct.error:
+            logger.debug(f"unpack message type failed {peer_response}")
+    return None, None
+
+
+
+async def peer_connection(peer, torrent_file, piece_manager):
+    try:
+        reader, writer = await asyncio.open_connection(peer.ip, peer.port)
+    except OSError:
+        logger.debug("connection call failed")
+        return False
+    logger.debug(f'{peer} connection opend')
+    if not await handshake(reader, writer, torrent_file):
+        logger.debug("connection failed")
+        return False
+    logger.debug(f"{peer} -  handshake sent")
+    if not await bitfield(reader, peer):
+        logger.debug(f"{peer} - bitfield failed")
+    logger.debug(f"{peer} - bitfiled success")
+    message_id = await intersted(reader, writer)
+    if message_id is None or message_id not in (0, 1):
+        logger.debug(f"{peer} - intersted request failed")
```

Comment:
don't you want to return in this case? 
in general you can split this function into lots of little functions for readability

Reply:
fixed

Reply:
all this process rebuild in next verison